### DISCUSSION TITLE
Plugin interface ideas

### DISCUSSION THREAD

**Discussion by jacksmith15** [jacksmith15/poetry-workspace-plugin#1] [Ideas]
This plugin is currently **experimental**, meaning that the CLI interface and behaviour is far from set in stone.  If you have any ideas about how it could be better,  please discuss here!

**Comment by cognifloyd** [comment]
Do you have any examples showing what a repo using those plugin looks like?
Do the subprojects have their own pyproject.toml or does everything get embedded in the root pyproject.toml?

**Reply by jacksmith15** [reply]
Yep sure - see [this branch](https://github.com/jacksmith15/poetry-workspace-plugin/tree/example/simple-workspaces)

In that example the root level [pyproject.toml](https://github.com/jacksmith15/poetry-workspace-plugin/blob/example/simple-workspaces/pyproject.toml#L59-L60) tracks two workspaces, both under [libs/](https://github.com/jacksmith15/poetry-workspace-plugin/tree/example/simple-workspaces/libs). One of those librarys ([my-library](https://github.com/jacksmith15/poetry-workspace-plugin/tree/example/simple-workspaces/libs/my-library)) depends on the other ([my-other-library](https://github.com/jacksmith15/poetry-workspace-plugin/tree/example/simple-workspaces/libs/my-other-library)).

The subprojects have their own `pyproject.toml` files. The root `pyproject.toml` just contains a mapping of name to path for each workspace.  This way its easy to add an existing poetry project as a workspace.

I like the idea that the workspaces themselves don't need to have any notion of being a workspace, and taken on their own they are just vanilla poetry projects.

**Reply by cognifloyd** [reply]
Cool. I'll check that link out in a moment.

Is there any coordination of dependencies? Ie something that makes sure the dependencies don't conflict with each other across workspaces?

**Reply by cognifloyd** [reply]
Oh. So the only way to make sure dependencies don't conflict is to make the workspaces depend on each other, right?

**Reply by jacksmith15** [reply]
That's currently the default behaviour, but there's nothing stopping this from being enforced by adding each workspace as a dependency to the root project.

Perhaps there should be an option to automatically have the root project depend on its workspaces - as I think that's definitely a desired use case (and indeed is how yarn workspaces work).

The reason I didn't want to enforce _all_ dependencies to be compatible is that you might have a set of build target workspaces which don't need compatible dependencies between each other, just with their dependencies.

For example:
1. I have two applications (A1, A2) which both depend on a shared library L1
2. Both A1 and A2 use a third-party dependency TP, but a require different versions

In this example, you still get the advantages of e.g. a single PR to update all three workspaces, test and build them.

**Reply by adriangb** [reply]
Just adding the workspaces as dependencies in the root package isn't the full picture. That will let you generate a lock file for the root package that includes all dependencies, but there's no way to install dependencies for a single workspace while respecting the versions in the lock file. It seems like the plug-in would have to do some magic to generate lock files for each workspace that are all in sync with the each other.

**Reply by jacksmith15** [reply]
A single lockfile would be a considerable usability improvement for many cases, but the current `poetry.lock` format doesn't allow aliasing multiple versions of the same dependency, and so it's a limited approach.

I believe a single lock file is do-able, by overriding the install logic for each subproject to only install a subset of the lockfile. I've also done some experiments with generating the subproject venvs from the root environment using symlinks to reduce disk usage/install time whilst keeping them isolated.

One other consideration for a shared lockfile, is whether poetry's pyproject dependencies structure would then make sense in a subproject. The version specifier becomes redundant, because all the subproject needs is a reference to the dependency in the lockfile.

**Reply by adriangb** [reply]
> the current poetry.lock format doesn't allow aliasing multiple versions of the same dependency, and so it's a limited approach

While it would be nice to be able to have multiple versions of a dependency for the reason you described above, it seems to work fine for Cargo and many large Python organizations. I've managed largo monorepos that do this and whenever there's been dependency conflicts it's been a smell of bad code or dependency bloat.

> I believe a single lock file is do-able, by overriding the install logic for each subproject to only install a subset of the lockfile

I was thinking the flow would be collect all dependencies -> make a top level lockfile -> for each package create it's own lockfile using the versions from the global lockfile. So that each package remains a valid poetry project with a lock file and all, just that those lock files are synced. I think this would just require hooking into poetry's machinery for reading dependencies from a pyproject.toml, the solver and writing out lock files.

> One other consideration for a shared lockfile, is whether poetry's pyproject dependencies structure would then make sense in a subproject. The version specifier becomes redundant, because all the subproject needs is a reference to the dependency in the lockfile.

It's not irrelevant: the package pyproject.toml files are the source of version constraints. For example one package might have foo>1.2.1 and another might have foo==1.4.3 which solves or foo==1.2.0 which should error. This is the same thing Cargo does.
